### PR TITLE
Made the map zoom work with the UN region filter

### LINKED ISSUE #116: Add filtering based on UN subregions
We want to add a filter that will allow users to filter on WHO subregions as well. Similar to WHO region when a subregion is selected we want the map to zoom in on that subregion. 

We want to make the subregion filter responsive to changes in the WHO region and country filters and vice versa.
- i.e. If we select a specific country, then the other 2 dropdowns only show options in which that country belong etc. Ask me for more details if confused. 

This should be additive. i.e. if multiple regions are selected the map zoom should include all the selected regions. 

https://www.google.com/url?q=https://unstats.un.org/unsd/methodology/m49/&sa=D&source=docs&ust=1704240536012873&usg=AOvVaw0kjgxUQNJsRAAgnRGNTTdG

### PR DESCRIPTION
**Pull request by SeanKennyNF**
Resolves #116. This makes the map zoom focus on the UN region when you pick one. I left a discussion in #116 about how the filtering with country and UN region doesn't play super well together but I don't think that's appropriate to hit as a part of this ticket since that's an issue already present with countries and WHO regions.

Related PRs in release order
- https://github.com/SeanKennyNF/iit-backend-v2/pull/4
- https://github.com/PathoTracker/Pathotracker/pull/126
- https://github.com/PathoTracker/Pathotracker/pull/127

Peek 2024-01-12 12-43

### CODE REVIEW

File: lib/bounding-boxes.ts [hunk 1]
```diff
@@ -196,31 +222,214 @@ const countryBoundingBoxData: Record<string, CountryBoundingBox | undefined> = {
   "Saint Lucia": [-61.07, 13.72, -60.85, 14.1],
 };
 
+const alphaTwoCountryCodeBoundingBoxData: Record<
```
Comment (SeanKennyNF) [inline lib/bounding-boxes.ts:225]: Made this because I don't want country names changing to break anything anymore :sob:

Ideally, we switch the country stuff to use this one as well and then delete the country one entirely. I'd love to do that and that was kind of my secret motive for adding `alphaTwoCountryCode` on the API but one thing at a time.

File: lib/bounding-boxes.ts [hunk 2]
```diff
@@ -196,31 +222,214 @@ const countryBoundingBoxData: Record<string, CountryBoundingBox | undefined> = {
...
+  typedObjectFromEntries(
+    Object.values(UNRegion).map((unRegion) => [
+      unRegion,
+      combineBoundingBoxes(
+        getCountryCodesInUNRegion({ unRegion })
+          .map((countryCode) => alphaTwoCountryCodeBoundingBoxData[countryCode])
+          .filter(
+            <T>(boundingBox: T | undefined): boundingBox is T => !!boundingBox
+          )
+      ),
```
Comment (SeanKennyNF) [inline lib/bounding-boxes.ts:429]: :heart: :heart: :heart:

File: lib/un-regions.ts [hunk 3]
```diff
@@ -21,4 +21,263 @@ export enum UNRegion {
   MELANESIA = "MELANESIA",
   MICRONESIA = "MICRONESIA",
   POLYNESIA = "POLYNESIA",
-}
\ No newline at end of file
+}
+
+const countryAlphaTwoCodeToUNRegionsMap: Partial<Record<string, UNRegion>> = {
```
Comment (SeanKennyNF) [inline lib/un-regions.ts:26]: Little bit of copy-pastrami from the backend

File: utils/typed-library-functions.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export const typedObjectFromEntries = <TKey extends string, TValue>(
+  input: [TKey, TValue][]
+): Record<TKey, TValue> => {
+  return Object.fromEntries(input) as Record<TKey, TValue>;
+};
```
Comment (SeanKennyNF) [inline utils/typed-library-functions.ts:5]: I cry every time :smiling_face_with_tear:

File: contexts/arbo-context.tsx [hunk 5]
```diff
@@ -157,27 +156,65 @@ export const arboReducer = (
...
 };
 
-const adjustMapPositionIfCountryFilterHasChanged = (
-  action: ArboAction,
-  map: MapRef
+interface GetAllBoundingBoxesFromExistingFiltersInput {
+  selectedFilters: Record<string, string[] | undefined>;
+}
+
+const getAllBoundingBoxesFromExistingFilters = (
```
Comment (SeanKennyNF) [inline contexts/arbo-context.tsx:163]: This part might be kind of hard to follow but basically the old code was pretty hacky and made an assumption that `country` was the only filter that could move the map. I need to look at the added filter as well as any existing filters now because there are two filters that determine the bounding box the map uses now. This isn't perfect looking either but yeah.

Q: Sean isn't this silly? I mean if you pick region "Eastern Europe" and "India" you'll filter out literally everything. Who cares if the map zooms in on the correct bounding box?

A: Yes you're right! This I left a discussion point on #116. I'd like to make a separate ticket to make any filters on geographical regions additive with one another so that "Eastern Europe" + "India" doesn't mean get me any "Indian  studies in Eastern Europe" and instead means "get me any studies from India or Eastern Europe".

**Review by EliasPinno — APPROVED** [review]
